### PR TITLE
Use connection URI in SqliteHook

### PR DESCRIPTION
**Pull request by feluelle**
We use the uri syntax for connecting to a sqlite database.
This allows the user to define more sqlite args such as mode. See https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#uri-connections for details.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/sqlite/hooks/sqlite.py
Comment (dstandish) [inline airflow/providers/sqlite/hooks/sqlite.py]: i assume this changes the output. will it break behavior for users?

Comment (dstandish) [reply airflow/providers/sqlite/hooks/sqlite.py]: i guess since it is a provider, we can break things, but you'll have to make it major release... i think the way to signal that to release manager is just add to changelog ?? @eladkal

Comment (eladkal) [reply airflow/providers/sqlite/hooks/sqlite.py]: If we can't deprecate first and must do a major release then we need to update provider.yaml major version to 4.0.0 and add relevant entry in the provider change log as part of this PR.

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py]: No, it does not change the output. We have tests which validates that [here](https://github.com/apache/airflow/pull/28721/files#diff-3614936e10a5d7aa9916f548b547cf3443f8ea29e30558d6fe55176e0a6f702eR30).

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py]: See line 139-140.

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py:49]: But what changes is what I mentioned in https://github.com/apache/airflow/pull/28721#issuecomment-1370717908

**Comment by feluelle** [comment]
I have found a solution  https://github.com/apache/airflow/pull/28721/commits/f3eaec7875fc7bc373069320cf72d5852dbbc812 which does not require the conn_type to be added on user side. PTAL @dstandish @eladkal 🙏

File: airflow/providers/sqlite/hooks/sqlite.py
Comment (eladkal) [reply airflow/providers/sqlite/hooks/sqlite.py]: breaking change can also be about changing the input schema.
I did not review the PR, I just commented that if this is a breaking change then just need adding entry in change log and updating provider.yaml if this is not a breaking change but just a bug fix then no further action is required.

**Comment by dstandish** [comment]
> I have found a solution [f3eaec7](https://github.com/apache/airflow/commit/f3eaec7875fc7bc373069320cf72d5852dbbc812) which does not require the conn_type to be added on user side. PTAL @dstandish @eladkal 🙏

nice, good thinking

Ok, so, it looks right to escape for compatibility with Connection.  But it does appear it will change if hostname happens to have url eschape codes in it.  but, to me this is like not a normal thing to do so, probably fine to ignore.   

however, i did look and Connection.from_uri does not escape the `host`... so we would be inconsistent with its parsing of uri in that respect. again probably maybe ok?  

But i did look to see what postgres does, and it doesn't escape.  Maybe sqlalchemy handles the escaping?  Can you confirm that before we merge?  If it's good enough for postgres,  maybe we don't need it here? It is probably worth adding a single test for URI behavior and a single one verifying that `sqlite3.connect` is called with the right value.

**Review by ashb — APPROVED** [review]
LGTM. Does it need docs https://airflow.apache.org/docs/apache-airflow-providers-sqlite/stable/connections/sqlite.html updating?

**Comment by josh-fell** [comment]
> LGTM. Does it need docs https://airflow.apache.org/docs/apache-airflow-providers-sqlite/stable/connections/sqlite.html updating?

+1. [There is a PR in-flight](https://github.com/apache/airflow/pull/29139#pullrequestreview-1268619921) which was planning to update the docs around connecting to SQLite in an attempt [to fix this issue](https://github.com/apache/airflow/issues/29113). Not sure how we typical handle this collision. Doing a single update in a commit would be best.

**Review by dstandish — APPROVED** [review]

**Comment by feluelle** [comment]
Yes, good catch @ashb and @josh-fell thank you for the heads-up. That is a tricky one. I would say we merge the other PR first and I will just update my PR and docs to reflect the changes i.e. that from now on you could use certain fields such as login, password, etc. The current documentation also states that the fields are required, which they are not even with my change using the URL.

But I would like to bring @atharvmishra1999 to the discussion. How do you think we should handle the collision?

File: airflow/providers/sqlite/hooks/sqlite.py
Comment (uranusjr) [inline airflow/providers/sqlite/hooks/sqlite.py]: Is `:memory:` a viable value here? This logic does not support that.

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py]: Yes, it does. For example, see the last example [here](https://docs.python.org/3/library/sqlite3.html#how-to-work-with-sqlite-uris).
> db = "file:mem1?mode=memory&cache=shared"
> con1 = sqlite3.connect(db, uri=True)

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py]: More examples can be found [here](https://www.sqlite.org/inmemorydb.html).

Comment (uranusjr) [reply airflow/providers/sqlite/hooks/sqlite.py]: So you can’t use the `"sqlite://"` string (the default way to create an in-memory SQLite database in SQLAlchemy) since no way to configure a connection that formats into this specific string, but the same effect can be achieved using the URL format, right? I wonder if we want to document this somewhere.

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py]: You can. The host field is optional. You can create an empty SQLite connection and it will use an in-memory db.

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py]: I have added a test case for that 37d2a53.

And validated that it works locally:
```
>>> import sqlite3
>>> sqlite3.connect("file:")
<sqlite3.Connection object at 0xffffb71a4d50>
```

File: airflow/providers/sqlite/hooks/sqlite.py
Comment (feluelle) [inline airflow/providers/sqlite/hooks/sqlite.py]: FYI: This is also compatible with in-memory DB's in sqlalchemy.

For example:
```
>>> import sqlalchemy
>>> sqlalchemy.create_engine('sqlite:///').connect()
<sqlalchemy.engine.base.Connection object at 0xffffa95e5580>
```

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (ashb) [inline docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: Host doesn't make any sense for a sqlite example

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (ashb) [inline docs/apache-airflow-providers-sqlite/connections/sqlite.rst:56]: What changed with 2.3? (I don't know what I was a user am meant to take from this paragraph)

File: tests/providers/sqlite/hooks/test_sqlite.py
Comment (ashb) [inline tests/providers/sqlite/hooks/test_sqlite.py:38]: Nit:
```suggestion
        ["connection", "uri"],
```

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst:56]: This is a copy&paste from the snowflake one.

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: This is the path to the database. In this example the file "host" will be created in the current directory.

Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: What do you want me to put there instead?

File: tests/providers/sqlite/hooks/test_sqlite.py
Comment (feluelle) [reply tests/providers/sqlite/hooks/test_sqlite.py:38]: This is not necessary.

Do you find your suggestion more readable?

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (ashb) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: Oh, have it as something like `sqlite:///path/to/db` then. And if it should only be `//` not `///` I think it's worth calling that out explicitly (as anyone familiar with sqlite might instinctively go for `///`)

Comment (ashb) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: That said... Why are we changing it away from sqlite norms? Why can't we have `AIRFLOW_CONN_SQLITE_DEFAULT=sqlite:///path/to/db`? (Sorry, I've lost any and all context to this PR of late)

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (ashb) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst:56]: Delete it then? It doesn't make any sense to me

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: 🤔 good question actually.. I will check tomorrow if we can remove the `replace` in `get_uri`.

Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: I think it is because of `get_uri` impl in connection. It always has two slashes. So we need to add another one. Even if we add an absolute path in the host field we are still missing one. The absolute path in sql alchemy requires four slashes in total.

tldr; The URI you see in the docs is a connection URI - not a sqlalchemy uri.

Comment (ashb) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: Thinking about it -- how does your PR behave if you try to specify a path to an absolute location? How would you create a URI for a db in `/tmp/my.sqlite`?

Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: `sqlite:///tmp/my.sqlite`

Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: You basically set the host field to the path and it will add another slash when getting the uri used for sql alchemy uris. For SQLite uris it will use `file:` instead.

Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst]: Since this is quite confusing, I will add an example with a relative path and one with an absolute path.

File: tests/providers/sqlite/hooks/test_sqlite.py
Comment (Taragolis) [reply tests/providers/sqlite/hooks/test_sqlite.py:38]: Yep, it not required for `pytest`, and in general we do not use it in our codebase

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (Taragolis) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst:56]: JSON format added: https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#json-format-example

So better clarify or remove it

**Comment by feluelle** [comment]
Actually, I have found some issues. Will hopefully resolve them on Monday.

File: docs/apache-airflow-providers-sqlite/connections/sqlite.rst
Comment (feluelle) [reply docs/apache-airflow-providers-sqlite/connections/sqlite.rst:56]: I have clarified it. :)

**Comment by feluelle** [comment]
The issues should all be resolved. I know that it does not look so great using string replacements, but I think it is safe to use it in these cases.

File: airflow/providers/sqlite/hooks/sqlite.py
Comment (feluelle) [inline airflow/providers/sqlite/hooks/sqlite.py]: The trailing slash in the path is a side effect of using `airflow_conn.get_uri()` 😞 I can also move the replace on this function call instead.

Comment (uranusjr) [reply airflow/providers/sqlite/hooks/sqlite.py]: It may be a good idea to move the logic to `get_uri`, this is arguably a bug.

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py]: I am not quite sure, because the URI format of an airflow connection is different from the one [SQLite uses](https://www.sqlite.org/uri.html).

Comment (feluelle) [reply airflow/providers/sqlite/hooks/sqlite.py]: SQLite `file:` URIs can't end with a slash because they are supposed to be files - not directories, but an airflow connection path can be anything.

**Review by uranusjr — APPROVED** [review]

**Pull request merged**
